sklearn/svm/src/libsvm/svm.cpp

```diff
@@ -1816,6 +1816,118 @@ static void solve_nu_svr(
 	delete[] y;
 }
 
+static void solve_svdd_l1(
+	const PREFIX(problem) *prob, const svm_parameter *param,
+	double *alpha, Solver::SolutionInfo* si)
+{
+	int l = prob->l;
+	int i, j;
+
+	double r_square;
+
+	ONE_CLASS_Q Q = ONE_CLASS_Q(*prob, *param);
+
+	if(param->nu < 1) {
+		// case \nu < 1: the dual problem is
+		//	min 0.5(\alpha^T Q \alpha) + (-0.5 \nu W diag Q)^T \alpha
+		//		e^T \alpha = \nu W
+		//		0 <= alpha_i <= W_i
+		//  W = sum W_i
+
+		schar *ones = new schar[l];
+		double *QD = new double[l];
+		double *linear_term = new double[l];
+			double *C = new double[l];
+
+		double nu_W = 0;
+		for(i=0;i<l;i++)
+		{
+			C[i] = prob->W[i];
+			nu_W += C[i] * param->nu;
+		}
+
+		for(i=0;i<l;i++)
+		{
+			QD[i] = nu_W * Q.get_QD()[i];
+			linear_term[i] = - QD[i] / 2;
+		}
+
+		for(i=0;i<l;i++)
+			ones[i] = 1;
+
+		double sum_alpha = nu_W;
+		i = 0;
+		while(sum_alpha > 0)
+		{
+			alpha[i] = min(C[i], sum_alpha);
+			sum_alpha -= alpha[i];
+			++i;
+		}
+		for(;i<l;i++)
+			alpha[i] = 0;
+
+		Solver s;
+		s.Solve(l, Q, linear_term, ones, alpha, C, param->eps,
+			si, param->shrinking, param->max_iter);
+
+		// Compute R: the solver returns
+		//  obj = 0.5 \alpha^T Q \alpha - 0.5 \nu W sum_i K_{ii}*\alpha_i
+		//  rho = 0.5 \nu W (\alpha^T Q \alpha / (\nu W)^2 - R)
+		r_square = 2*(si->obj - nu_W * si->rho);
+		for(i=0;i<l;i++)
+			r_square += alpha[i] * QD[i];
+		r_square /= nu_W * nu_W;
+
+	        delete[] C;
+		delete[] linear_term;
+		delete[] QD;
+		delete[] ones;
+	} else {
+		// case \nu >= 1: then R = 0, and the SVDD-L1 problem is reduced to
+		//  a quadratic problem with a unique solution independent of \nu.
+		// The centre of the sphere is the average of feature maps with weights W_i.
+
+		info("*\nSVDD-L1 solution independent of nu\n");
+
+		double sum_W = 0;
+		for(i=0;i<l;i++)
+		{
+			alpha[i] = prob->W[i];
+			si->upper_bound[i] = prob->W[i];
+			sum_W += prob->W[i];
+		}
+
+		// Simulate the run of the Solver by computing the objective
+		//  and the intercept:
+		//    obj = 0.5 \alpha^T Q \alpha - 0.5 W sum_i K_{ii}*\alpha_i
+		//    rho = 0.5 \alpha^T Q \alpha / W
+		// note that \sum_i \alpha_i = W.
+		double rho = 0;
+		double obj = 0;
+		double sum;
+		for(i=0;i<l;i++)
+		{
+			const Qfloat *Q_i = Q.get_Q(i,l);
+
+			obj -= sum_W * alpha[i] * Q_i[i] / 2;
+
+			// Utilize the symmetry of Q
+			sum = alpha[i] * Q_i[i] / 2;
+			for(j=i+1;j<l;j++)
+				sum += alpha[j] * Q_i[j];
+			rho += alpha[i] * sum;
+		}
+		si->obj = rho + obj;
+		si->rho = rho / sum_W;
+
+		si->solve_timed_out = false;
+
+		r_square = 0.0;
+	}
+
+	info("R^2 = %f\n",r_square);
+}
+
```

Comment:
Is this implementation derived from upstream libsvm? Are there scikit-learn specific changes in it? If so please add a comment to to give an overview of what is specific to the scikit-learn version.